packages/next/shared/lib/router/router.ts

```diff
@@ -357,7 +357,10 @@ export function resolveHref(
       finalUrl.searchParams &&
       resolveAs
     ) {
-      const query = searchParamsToUrlQuery(finalUrl.searchParams)
+      let query = searchParamsToUrlQuery(finalUrl.searchParams)
+      if (typeof href !== 'string') {
+        query = { ...query, ...(href.query as ParsedUrlQuery) }
```

Comment:
The only place I could find `query` to be a `string` is the (I believe) now-deprecated serverless loader https://github.com/vercel/next.js/blob/352e60deddab9667d045b4eca2a77ef25a8adbaa/packages/next/build/webpack/loaders/next-serverless-loader/page-handler.ts#L247

But let me know if I overlook something :thinking: